### PR TITLE
91 laplace ggn arguments

### LINKED ISSUE #91: More flexible Laplace GGN arguments
Currently if you try to use e.g. `laplace.dense_ggn` with
```python
def forward(p, b):
    x, _ = b
    return torch.func.functional_call(model, p, x), torch.tensor([])


def outer_log_lik(y_pred, b):
    _, y = b
    return torch.distributions.Normal(y_pred, obs_sd).log_prob(y).sum()
```

It will throw and error
```
TypeError: forward() got an unexpected keyword argument 'batch'
```

We can fix this with a `lambda` rather than `partial`

### PR DESCRIPTION
**Pull request by BenjaminDev**
Thanks for the very interesting library. I generally like to attempt fixing issues to get to know a code base. Feel free to just close this PR if this change is missing the point. 

Fixes #91

### CODE REVIEW

File: posteriors/laplace/dense_ggn.py [hunk 1]
```diff
@@ -137,7 +137,7 @@ def outer_loss(z, batch):
 
     with torch.no_grad(), CatchAuxError():
         ggn_batch, aux = ggn(
-            partial(forward, batch=batch),
+            lambda params: forward(params, batch),
             partial(outer_loss, batch=batch),
```
Comment (SamDuffield) [inline posteriors/laplace/dense_ggn.py:141]: Think we need to change `outer_loss` to a `lambda` too

File: tests/laplace/test_dense_ggn.py [hunk 2]
```diff
@@ -18,8 +18,8 @@ def normal_log_likelihood(y_pred, batch):
     )  # validate args introduces control flows not yet supported in torch.func.vmap
 
 
-def forward_m(params, batch, model):
-    y_pred = functional_call(model, params, batch[0])
+def forward_m(params, b, model):
+    y_pred = functional_call(model, params, b[0])
```
Comment (SamDuffield) [inline tests/laplace/test_dense_ggn.py:22]: Not sure this modification is needed 😄

Comment (SamDuffield) [reply tests/laplace/test_dense_ggn.py:22]: Actually never mind me! I see this is now testing the argument not being called `batch` - good stuff! 😸

File: posteriors/laplace/dense_ggn.py (hunk 1 above)
Comment (BenjaminDev) [reply posteriors/laplace/dense_ggn.py]: It's not strictly needed as the `outer_loss` function is defined above `L135` (by the lib) and so kw arguments pass just fine.

Updated to use a lambda. I think having them the same is simpler to read.

**Comment by SamDuffield** [comment]
Looking good! Can we do the same for `laplace.diag_ggn` too? Then we should be good to merge

**Comment by BenjaminDev** [comment]
@SamDuffield, Just one comment after thinking this change over. `Lambda`'s in python are late binding which means this change from `partial` to `lambda x: ...` is subtly different.

To show the difference consider:
```
with torch.no_grad(), CatchAuxError():
        forward_partial = partial(forward, batch=batch) # 
        diag_ggn_fn = diag_ggn(
            forward_partial, # Original pattern which forces consumers of the lib to use the `batch` kw arg. But! batch is bound on this line.
            lambda z: outer_loss(z, batch), # New pattern BUT! `batch` is only bound when the the lambda is invoked. i.e when `diag_ggn_fn` is called.
            forward_has_aux=True,
            loss_has_aux=False,
            normalize=False,
        )
        batch=None # If batch is modified the `partial` function is unaffected as it bound to `batch` in `forward_partial.keywords['batch']`. But the `lambda` only binds later and will be affected. 
        diag_ggn_batch, aux = ff(state.params)
``` 

A pattern that makes a `lambda` bind early is:

```
    ...
     diag_ggn(
            lambda params, _batch=batch: forward(params, _batch)# Now `batch` binds in the  `.__defaults__` 
            ...
    )
 ...
``` 


I updated the diff to reflect the way we were both heading and I think it's a perfectly fine solution and all tests pass. But, I just wanted to point out that my small change is, on a "python" level significantly different.

I think we good to merge unless the above trigger larger concern on your side.

**Review by SamDuffield — APPROVED** [review]
That is a very good point, but as you note I think all is good as we enforce the specific (params, batch) signature. So good to merge!

**Pull request merged**